providers/implementations/kem/ml_kem.c

```diff
@@ -19,214 +19,241 @@
-        return 0;
+    if (*secretlen < ML_KEM_SHARED_SECRET_BYTES) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH,
+                       "short ML-KEM encapsulate shared secret");
+        goto end;
+    }
+    if (*outlen < v->ctext_bytes) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH,
+                       "short ML-KEM encapsulate ciphertext");
+        goto end;
+    }
 
-    if (ctx->entropy != NULL) {
-        ret = ossl_mlkem768_encap_external_entropy(out, secret,
-                                                   &ctx->key->pubkey,
-                                                   ctx->entropy,
-                                                   ctx->key->mlkem_ctx);
-    } else {
-        ret = ossl_mlkem768_encap(out, (uint8_t *)secret, &ctx->key->pubkey,
-                                  ctx->key->mlkem_ctx);
+    if (secret == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_SECRET);
+        goto end;
     }
 
-    debug_print("MLKEMKEM encaps returns %d\n", ret);
+    *secretlen = ML_KEM_SHARED_SECRET_BYTES;
+    *outlen = key->vinfo->ctext_bytes;
+    if (ctx->entropy != NULL)
+        ret = ossl_ml_kem_vencap_seed(v, out, *outlen, secret, *secretlen,
+                                      key->pubkey, key->prvkey, ctx->entropy,
+                                      ML_KEM_RANDOM_BYTES, key->ctx);
+    else
+        ret = ossl_ml_kem_vencap_rand(v, out, *outlen, secret, *secretlen,
+                                      key->pubkey, key->prvkey, key->ctx);
+
+  end:
+    if (ctx->entropy != NULL) {
+        OPENSSL_cleanse(ctx->entropy, ML_KEM_RANDOM_BYTES);
+        ctx->entropy = NULL;
+    }
     return ret;
 }
 
-static int mlkem_decapsulate(void *vctx, unsigned char *out, size_t *outlen,
-                             const unsigned char *in, size_t inlen)
+static int ml_kem_decapsulate(void *vctx, unsigned char *out, size_t *outlen,
+                              const unsigned char *in, size_t inlen)
 {
-    PROV_MLKEM_CTX *ctx = (PROV_MLKEM_CTX *)vctx;
-    int ret;
+    PROV_ML_KEM_CTX *ctx = vctx;
+    ML_KEM_PROVIDER_KEYPAIR *key = ctx->key;
+    vinfo_t v;
 
-    debug_print("MLKEMKEM decaps %p to %p\n", ctx, out);
-    debug_print("MLKEMKEM decaps inlen at %ld\n", inlen);
-    if (outlen != NULL)
-        *outlen = OSSL_MLKEM768_SHARED_SECRET_BYTES;
+    if (key == NULL || key->prvkey == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
+        return 0;
+    }
+    v = key->vinfo;
 
     if (out == NULL) {
-        debug_print("MLKEMKEM decaps outlen set to %ld \n", *outlen);
+        if (outlen == NULL)
+            return 0;
+        *outlen = ML_KEM_SHARED_SECRET_BYTES;
         return 1;
     }
 
-    if (ctx->key == NULL
-        || ctx->key->keytype != MLKEM_KEY_TYPE_768
-        || ctx->key->encoded_privkey == NULL
-        || in == NULL)
-        return 0;
-
-    if (inlen != OSSL_MLKEM768_CIPHERTEXT_BYTES)
+    if (*outlen < ML_KEM_SHARED_SECRET_BYTES) {
```

Comment:
Should we check `outlen != NULL` here? We do in the `out == NULL` branch above - but not here?

Reply:
I don't know, you know more about provide internals, is there a way to convince EVP to send NULL output length pointers to the encap/decap provider implementation?

Reply:
I had thought it unlikely/not useful for`outlen` to be NULL, while `out` is not.
What then is the length of the output buffer (shared secret result) the provider is supposed to write to?

However, it looks like the EVP layer does not preclude this possibility: https://github.com/openssl/openssl/blob/e0ea913f11cf64d000556bbf7cb9f8acdf6be4cb/crypto/evp/kem.c#L256-L259
The length pointer is only enforced non-NULL when `out` **is** NULL.

The documentation states: https://github.com/openssl/openssl/blob/e0ea913f11cf64d000556bbf7cb9f8acdf6be4cb/doc/man3/EVP_PKEY_decapsulate.pod?plain=1#L35-L40

So it seems that perhaps callers are allowed to *promise* they've allocated enough space, and not indicate how much space that was.

In which case, indeed I need to be able to handle this, by substituting the address of a `size_t` with the required value when the provided pointer is NULL.